File: ai-engine/src/itinerary/stabilityEngine.js
import { calculateConfidence } from "../intelligence/confidenceScoring.js";
import { clamp, normalizeText } from "../retrieval/textUtils.js";
import {
  replanDeterministicItinerary,
  validateDeterministicItinerary
} from "./deterministicPlanner.js";

export function replanWithStability({
  previousItinerary,
  itineraryInput = {},
  changes = {},
  places,
  stabilityOptions = {}
} = {}) {
  const nextItinerary = replanDeterministicItinerary({
    itineraryInput,
    changes,
    places
  });

  return stabilizeItinerary({
    previousItinerary,
    nextItinerary,
    changes,
    ...stabilityOptions
  });
}

export function stabilizeItinerary({
  previousItinerary,
  nextItinerary,
  changes = {},
  lockedActivityIds = [],
  acceptedActivityIds = [],
  preserveAccepted = true,
  forceFreshDays = []
} = {}) {
  if (!previousItinerary?.days?.length || !nextItinerary?.days?.length) {
    const stabilityReport = buildEmptyStabilityReport(previousItinerary, nextItinerary);
    return {
      ...nextItinerary,
      stage: "stable-itinerary-replan-v1",
      stabilityReport,
      ...calculateConfidence({
        base: nextItinerary?.confidence || 0.72,
        outputCompleteness: nextItinerary?.days?.length ? 1 : 0.4
      })
    };
  }

  const affectedDays = inferAffectedDays(previousItinerary, changes, forceFreshDays);
  const lockedIds = new Set([...lockedActivityIds, ...acceptedActivityIds].map(String));
  const preservedDays = [];
  const regeneratedDays = [];
  const stableDays = nextItinerary.days.map((nextDay) => {
    const previousDay = findMatchingPreviousDay(previousItinerary.days, nextDay);
    const canPreserve = previousDay
      && !affectedDays.has(nextDay.day)
      && !hasBlockedActivity(previousDay, changes)
      && (preserveAccepted || !hasLockedActivity(previousDay, lockedIds));

    if (canPreserve) {
      preservedDays.push(nextDay.day);
      return recalculateDayTotals({
        ...previousDay,
        stability: {
          preserved: true,
          source: "previous-itinerary",
          reason: "No direct change affected this day, so accepted structure was preserved."
        }
      });
    }

    regeneratedDays.push(nextDay.day);
    return recalculateDayTotals({
      ...nextDay,
      stability: {
        preserved: false,
        source: "fresh-replan",
        reason: affectedDays.has(nextDay.day)
          ? "This day was directly affected by the change set."
          : "No compatible previous day was available."
      }
    });
  });
  const itinerary = {
    ...nextItinerary,
    stage: "stable-itinerary-replan-v1",
    generatedAt: new Date().toISOString(),
    days: stableDays,
    ...buildDayAliases(stableDays),
    totals: calculateTotals(stableDays)
  };
  const stabilityReport = buildStabilityReport({
    previousItinerary,
    itinerary,
    affectedDays,
    preservedDays,
    regeneratedDays,
    lockedActivityIds: [...lockedIds]
  });
  const feasibility = validateDeterministicItinerary({
    ...itinerary,
    constraints: nextItinerary.constraints
  });
  const confidenceReport = calculateConfidence({
    base: nextItinerary.confidence || 0.84,
    feasible: feasibility.valid,
    outputCompleteness: stabilityReport.previousActivityCount
      ? stabilityReport.currentActivityCount / stabilityReport.previousActivityCount
      : 1,
    fallbackWarnings: nextItinerary.confidenceReasons?.filter((reason) => reason.includes("fallback")) || []
  });

  return {
    ...itinerary,
    feasibility,
    stabilityReport,
    ...confidenceReport
  };
}

export function buildStabilityReport({
  previousItinerary,
  itinerary,
  affectedDays = new Set(),
  preservedDays = [],
  regeneratedDays = [],
  lockedActivityIds = []
} = {}) {
  const previousActivities = collectPlaceActivities(previousItinerary?.days || []);
  const currentActivities = collectPlaceActivities(itinerary?.days || []);
  const previousKeys = new Set(previousActivities.map((activity) => activity.key));
  const currentKeys = new Set(currentActivities.map((activity) => activity.key));
  const preservedActivities = currentActivities.filter((activity) => previousKeys.has(activity.key));
  const replacedActivities = previousActivities.filter((activity) => !currentKeys.has(activity.key));
  const newActivities = currentActivities.filter((activity) => !previousKeys.has(activity.key));
  const movedActivities = preservedActivities.filter((activity) => {
    const previous = previousActivities.find((item) => item.key === activity.key);
    return previous && previous.day !== activity.day;
  });
  const stabilityScore = previousActivities.length
    ? preservedActivities.length / previousActivities.length
    : 1;

  return {
    previousActivityCount: previousActivities.length,
    currentActivityCount: currentActivities.length,
    preservedActivityCount: preservedActivities.length,
    replacedActivityCount: replacedActivities.length,
    newActivityCount: newActivities.length,
    movedActivityCount: movedActivities.length,
    stabilityScore: Number(clamp(stabilityScore * 100, 0, 100).toFixed(2)),
    affectedDays: [...affectedDays].sort((a, b) => a - b),
    preservedDays,
    regeneratedDays,
    lockedActivityIds,
    preservedActivities: publicActivities(preservedActivities),
    replacedActivities: publicActivities(replacedActivities),
    newActivities: publicActivities(newActivities),
    movedActivities: publicActivities(movedActivities),
    rule: "Preserve unaffected accepted days; regenerate only directly impacted days."
  };
}

function buildEmptyStabilityReport(previousItinerary, nextItinerary) {
  return {
    previousActivityCount: collectPlaceActivities(previousItinerary?.days || []).length,
    currentActivityCount: collectPlaceActivities(nextItinerary?.days || []).length,
    preservedActivityCount: 0,
    replacedActivityCount: 0,
    newActivityCount: collectPlaceActivities(nextItinerary?.days || []).length,
    movedActivityCount: 0,
    stabilityScore: 0,
    affectedDays: [],
    preservedDays: [],
    regeneratedDays: (nextItinerary?.days || []).map((day) => day.day),
    lockedActivityIds: [],
    preservedActivities: [],
    replacedActivities: [],
    newActivities: publicActivities(collectPlaceActivities(nextItinerary?.days || [])),
    movedActivities: [],
    rule: "No previous itinerary was available, so no structure could be preserved."
  };
}

function inferAffectedDays(previousItinerary, changes, forceFreshDays = []) {
  const affectedDays = new Set(forceFreshDays.map(Number).filter(Boolean));
  const days = previousItinerary.days || [];

  for (const day of changes.affectedDays || []) affectedDays.add(Number(day));
  if (changes.day || changes.dayNumber) affectedDays.add(Number(changes.day || changes.dayNumber));
  if (changes.flightDelayHours || changes.flightDelayMinutes) affectedDays.add(1);
  for (const day of changes.skipDays || []) {
    const dayNumber = Number(day);
    if (dayNumber) {
      for (const existingDay of days) {
        if (existingDay.day >= dayNumber) affectedDays.add(existingDay.day);
      }
    }
  }

  const changedPlaceIds = new Set([
    ...(changes.blockedPlaceIds || []),
    ...Object.entries(changes.openingHours || {})
      .filter(([, value]) => value?.closed || value?.temporarilyClosed)
      .map(([id]) => id)
  ].map(String));
  for (const day of days) {
    if (day.activities?.some((activity) => changedPlaceIds.has(String(activity.id)))) {
      affectedDays.add(day.day);
    }
  }

  const weatherKeys = Object.keys(changes.weather || {});
  if (weatherKeys.includes("*") || normalizeText(changes.weather?.condition).includes("rain")) {
    days.forEach((day) => affectedDays.add(day.day));
  } else {
    for (const key of weatherKeys.map(normalizeText)) {
      for (const day of days) {
        if (normalizeText(day.routeLocation) === key) affectedDays.add(day.day);
      }
    }
  }

  if (changes.forceFullReplan) {
    days.forEach((day) => affectedDays.add(day.day));
  }

  return affectedDays;
}

function findMatchingPreviousDay(previousDays, nextDay) {
  return previousDays.find((day) => day.day === nextDay.day)
    || previousDays.find((day) => day.routeLocation && day.routeLocation === nextDay.routeLocation);
}

function hasBlockedActivity(day, changes) {
  const blockedIds = new Set([
    ...(changes.blockedPlaceIds || []),
    ...Object.entries(changes.openingHours || {})
      .filter(([, value]) => value?.closed || value?.temporarilyClosed)
      .map(([id]) => id)
  ].map(String));
  return day.activities?.some((activity) => blockedIds.has(String(activity.id)));
}

function hasLockedActivity(day, lockedIds) {
  if (!lockedIds.size) return false;
  return day.activities?.some((activity) => lockedIds.has(String(activity.id)));
}

function collectPlaceActivities(days) {
  return days.flatMap((day) => (day.activities || [])
    .filter((activity) => activity.kind === "place")
    .map((activity) => ({
      ...activity,
      day: day.day,
      key: activityKey(activity)
    })));
}

function activityKey(activity) {
  return String(activity.id || normalizeText(activity.name));
}

function publicActivities(activities) {
  return activities.map((activity) => ({
    id: activity.id || null,
    name: activity.name,
    day: activity.day,
    startTime: activity.startTime,
    endTime: activity.endTime
  }));
}

function recalculateDayTotals(day) {
  const placeActivities = (day.activities || []).filter((activity) => activity.kind === "place");
  return {
    ...day,
    totalCost: placeActivities.reduce((sum, activity) => sum + (activity.estimatedGroupCost || 0), 0),
    totalFatigue: Number(placeActivities.reduce((sum, activity) => sum + (activity.fatigueScore || 0), 0).toFixed(2)),
    totalWalkingHours: Number(placeActivities.reduce((sum, activity) => sum + (activity.walkingHours || 0), 0).toFixed(2))
  };
}

function calculateTotals(days) {
  return {
    totalCost: days.reduce((sum, day) => sum + (day.totalCost || 0), 0),
    totalFatigue: Number(days.reduce((sum, day) => sum + (day.totalFatigue || 0), 0).toFixed(2)),
    totalWalkingHours: Number(days.reduce((sum, day) => sum + (day.totalWalkingHours || 0), 0).toFixed(2)),
    scheduledPlaceCount: days.reduce(
      (sum, day) => sum + (day.activities || []).filter((activity) => activity.kind === "place").length,
      0
    )
  };
}

function buildDayAliases(days) {
  return Object.fromEntries(days.map((day) => [`day${day.day}`, day.activities]));
}

File: ai-engine/app/services/stability_engine.py
from datetime import datetime, timezone

from .confidence import calculate_confidence
from .itinerary_planner import calculate_totals, replan_deterministic_itinerary, validate_deterministic_itinerary
from .text import clamp, normalize_text


def replan_with_stability(
    *,
    previousItinerary: dict | None = None,
    previous_itinerary: dict | None = None,
    itineraryInput: dict | None = None,
    itinerary_input: dict | None = None,
    changes: dict | None = None,
    places: list[dict] | None = None,
    stabilityOptions: dict | None = None,
    stability_options: dict | None = None,
) -> dict:
    previous = previousItinerary or previous_itinerary
    next_itinerary = replan_deterministic_itinerary(
        itineraryInput=itineraryInput or itinerary_input or {},
        changes=changes or {},
        places=places,
    )
    return stabilize_itinerary(
        previousItinerary=previous,
        nextItinerary=next_itinerary,
        changes=changes or {},
        **(stabilityOptions or stability_options or {}),
    )


def stabilize_itinerary(
    *,
    previousItinerary: dict | None = None,
    previous_itinerary: dict | None = None,
    nextItinerary: dict | None = None,
    next_itinerary: dict | None = None,
    changes: dict | None = None,
    lockedActivityIds: list[str] | None = None,
    acceptedActivityIds: list[str] | None = None,
    preserveAccepted: bool = True,
    forceFreshDays: list[int] | None = None,
) -> dict:
    previous = previousItinerary or previous_itinerary
    next_plan = nextItinerary or next_itinerary
    changes = changes or {}
    if not previous or not previous.get("days") or not next_plan or not next_plan.get("days"):
        report = build_empty_stability_report(previous, next_plan)
        return {
            **(next_plan or {}),
            "stage": "stable-itinerary-replan-v1",
            "stabilityReport": report,
            **calculate_confidence(base=(next_plan or {}).get("confidence", 0.72), output_completeness=1 if (next_plan or {}).get("days") else 0.4),
        }

    affected_days = infer_affected_days(previous, changes, forceFreshDays or [])
    locked_ids = {str(value) for value in [*(lockedActivityIds or []), *(acceptedActivityIds or [])]}
    preserved_days: list[int] = []
    regenerated_days: list[int] = []
    stable_days: list[dict] = []
    for next_day in next_plan["days"]:
        previous_day = find_matching_previous_day(previous["days"], next_day)
        can_preserve = (
            previous_day
            and next_day["day"] not in affected_days
            and not has_blocked_activity(previous_day, changes)
            and (preserveAccepted or not has_locked_activity(previous_day, locked_ids))
        )
        if can_preserve:
            preserved_days.append(next_day["day"])
            stable_days.append(recalculate_day_totals({
                **previous_day,
                "stability": {
                    "preserved": True,
                    "source": "previous-itinerary",
                    "reason": "No direct change affected this day, so accepted structure was preserved.",
                },
            }))
        else:
            regenerated_days.append(next_day["day"])
            stable_days.append(recalculate_day_totals({
                **next_day,
                "stability": {
                    "preserved": False,
                    "source": "fresh-replan",
                    "reason": "This day was directly affected by the change set."
                    if next_day["day"] in affected_days
                    else "No compatible previous day was available.",
                },
            }))

    itinerary = {
        **next_plan,
        "stage": "stable-itinerary-replan-v1",
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "days": stable_days,
        **{f"day{day['day']}": day["activities"] for day in stable_days},
        "totals": calculate_totals(stable_days),
    }
    report = build_stability_report(
        previousItinerary=previous,
        itinerary=itinerary,
        affectedDays=affected_days,
        preservedDays=preserved_days,
        regeneratedDays=regenerated_days,
        lockedActivityIds=sorted(locked_ids),
    )
    feasibility = validate_deterministic_itinerary(itinerary)
    confidence = calculate_confidence(
        base=next_plan.get("confidence", 0.84),
        feasible=feasibility["valid"],
        output_completeness=report["currentActivityCount"] / report["previousActivityCount"] if report["previousActivityCount"] else 1,
    )
    return {**itinerary, "feasibility": feasibility, "stabilityReport": report, **confidence}


def build_stability_report(
    *,
    previousItinerary: dict | None = None,
    itinerary: dict | None = None,
    affectedDays: set[int] | None = None,
    preservedDays: list[int] | None = None,
    regeneratedDays: list[int] | None = None,
    lockedActivityIds: list[str] | None = None,
) -> dict:
    previous_activities = collect_place_activities((previousItinerary or {}).get("days", []))
    current_activities = collect_place_activities((itinerary or {}).get("days", []))
    previous_keys = {activity["key"] for activity in previous_activities}
    current_keys = {activity["key"] for activity in current_activities}
    preserved = [activity for activity in current_activities if activity["key"] in previous_keys]
    replaced = [activity for activity in previous_activities if activity["key"] not in current_keys]
    new = [activity for activity in current_activities if activity["key"] not in previous_keys]
    moved = [
        activity
        for activity in preserved
        if next((old for old in previous_activities if old["key"] == activity["key"]), {}).get("day") != activity["day"]
    ]
    stability_score = len(preserved) / len(previous_activities) if previous_activities else 1
    return {
        "previousActivityCount": len(previous_activities),
        "currentActivityCount": len(current_activities),
        "preservedActivityCount": len(preserved),
        "replacedActivityCount": len(replaced),
        "newActivityCount": len(new),
        "movedActivityCount": len(moved),
        "stabilityScore": round(clamp(stability_score * 100, 0, 100), 2),
        "affectedDays": sorted(affectedDays or set()),
        "preservedDays": preservedDays or [],
        "regeneratedDays": regeneratedDays or [],
        "lockedActivityIds": lockedActivityIds or [],
        "preservedActivities": public_activities(preserved),
        "replacedActivities": public_activities(replaced),
        "newActivities": public_activities(new),
        "movedActivities": public_activities(moved),
        "rule": "Preserve unaffected accepted days; regenerate only directly impacted days.",
    }


def build_empty_stability_report(previous, next_plan):
    return {
        "previousActivityCount": len(collect_place_activities((previous or {}).get("days", []))),
        "currentActivityCount": len(collect_place_activities((next_plan or {}).get("days", []))),
        "preservedActivityCount": 0,
        "replacedActivityCount": 0,
        "newActivityCount": len(collect_place_activities((next_plan or {}).get("days", []))),
        "movedActivityCount": 0,
        "stabilityScore": 0,
        "affectedDays": [],
        "preservedDays": [],
        "regeneratedDays": [day["day"] for day in (next_plan or {}).get("days", [])],
        "lockedActivityIds": [],
        "preservedActivities": [],
        "replacedActivities": [],
        "newActivities": public_activities(collect_place_activities((next_plan or {}).get("days", []))),
        "movedActivities": [],
        "rule": "No previous itinerary was available, so no structure could be preserved.",
    }


def infer_affected_days(previous: dict, changes: dict, force_fresh_days: list[int]) -> set[int]:
    affected = {int(day) for day in force_fresh_days if day}
    days = previous.get("days", [])
    for day in changes.get("affectedDays", []):
        affected.add(int(day))
    if changes.get("day") or changes.get("dayNumber"):
        affected.add(int(changes.get("day") or changes.get("dayNumber")))
    if changes.get("flightDelayHours") or changes.get("flightDelayMinutes"):
        affected.add(1)
    for day in changes.get("skipDays", []):
        day_number = int(day)
        for existing_day in days:
            if existing_day["day"] >= day_number:
                affected.add(existing_day["day"])
    changed_ids = {
        str(place_id)
        for place_id in [
            *(changes.get("blockedPlaceIds") or []),
            *[
                place_id
                for place_id, value in (changes.get("openingHours") or {}).items()
                if value.get("closed") or value.get("temporarilyClosed")
            ],
        ]
    }
    for day in days:
        if any(str(activity.get("id")) in changed_ids for activity in day.get("activities", [])):
            affected.add(day["day"])
    weather = changes.get("weather") or {}
    if "*" in weather or normalize_text(weather.get("condition") if isinstance(weather, dict) else weather).find("rain") >= 0:
        for day in days:
            affected.add(day["day"])
    if changes.get("forceFullReplan"):
        for day in days:
            affected.add(day["day"])
    return affected


def find_matching_previous_day(previous_days: list[dict], next_day: dict) -> dict | None:
    return next((day for day in previous_days if day["day"] == next_day["day"]), None) or next(
        (day for day in previous_days if day.get("routeLocation") and day.get("routeLocation") == next_day.get("routeLocation")),
        None,
    )


def has_blocked_activity(day: dict, changes: dict) -> bool:
    blocked_ids = {
        str(place_id)
        for place_id in [
            *(changes.get("blockedPlaceIds") or []),
            *[
                place_id
                for place_id, value in (changes.get("openingHours") or {}).items()
                if value.get("closed") or value.get("temporarilyClosed")
            ],
        ]
    }
    return any(str(activity.get("id")) in blocked_ids for activity in day.get("activities", []))


def has_locked_activity(day: dict, locked_ids: set[str]) -> bool:
    return bool(locked_ids) and any(str(activity.get("id")) in locked_ids for activity in day.get("activities", []))


def collect_place_activities(days: list[dict]) -> list[dict]:
    activities = []
    for day in days:
        for activity in day.get("activities", []):
            if activity.get("kind") == "place":
                key = str(activity.get("id") or normalize_text(activity.get("name")))
                activities.append({**activity, "day": day["day"], "key": key})
    return activities


def public_activities(activities: list[dict]) -> list[dict]:
    return [
        {
            "id": activity.get("id"),
            "name": activity.get("name"),
            "day": activity.get("day"),
            "startTime": activity.get("startTime"),
            "endTime": activity.get("endTime"),
        }
        for activity in activities
    ]


def recalculate_day_totals(day: dict) -> dict:
    places = [activity for activity in day.get("activities", []) if activity.get("kind") == "place"]
    return {
        **day,
        "totalCost": sum(activity.get("estimatedGroupCost", 0) for activity in places),
        "totalFatigue": round(sum(activity.get("fatigueScore", 0) for activity in places), 2),
        "totalWalkingHours": round(sum(activity.get("walkingHours", 0) for activity in places), 2),
    }
